### DISCUSSION TITLE
Browser back button and closing modals.

### DISCUSSION THREAD

**Discussion by ovenum** [unpoly/unpoly#355] [Q&A]
Hello! Just starting with unpoly,.js, so i might have missed something. 

Using the browser back button with an open modal or Layer will re-render the  document body. 
Is it possible to have the same behaviour as if `up.layer.dismiss` was called to close the modal? 
Re-rendering can cause flickering for lazy loaded images or other components that are reinitialized. 

**Example**
On the [unpoly.com](unpoly.com) frontpage most links in the main content section will open inside a modal. This modal can be dismissed by clicking the close icon in top right corner, but using the browsers back button to navigate back to the frontpage will close the modal _(good)_ AND re-render the complete body _(not what i want)_ causing the sidebar to flicker because it’s initialized again. 

Is there a way to intercept `up.render()` when its call originated from a `popstate` event and just calling `up.layer.dismiss` instead? 

`restoreStateOnPop` in history.js does emit _up:location:changed_ but only after up.render was called.

https://github.com/unpoly/unpoly/blob/b71e649082194e76aa0d8132ee5f334259b6563a/src/unpoly/history.js#L238-L267

**Comment by adam12** [comment]
I remember vaguely that this wasn't possible, but I haven't checked recently.

As a short-cut, can you `[up-keep]` your sidebar?

**Reply by ovenum** [reply]
HI Adam! Thanks for your reply! 
Using `[up-keep]`still seems to re-init JS components and loads images when using browser back button. This seems weird and not sure if i did use up-keep in the right way. 

Created a [simplified test on Glitch.com](https://glitch.com/edit/#!/erratic-boulder-airplane?path=index.html%3A21%3A5) illustrating the issue. It’s not using layers but renders would-be overlay content into another fragment.

```html
<div 
   x-data="SiteContent()"         
   id="site-content"         
   class="site-content"
   up-keep
>
  <!-- Content that should not be re-rendered -->
>

<div class="overlay">
  <!-- Overlay content goes here -->
</div>

```

Also noticed that calling `preventDefault` inside an `up:fragment:loaded` listener and closing overlays manually might work.

```js
/**
 * Avoid re-initialization of components or flickering when browser back button is
 * used while a modal is open.
 *
 * This should mimick the same behaviour as if up.layer.dismiss was called to
 * close the overlay.
 */
up.on("up:fragment:loaded", (event) => {
  const { renderOptions } = event;

  if (renderOptions?.baseLayer.mode === "cover") {
    // Prevent up.render from rendering fragment
    event.preventDefault();

    // Close all layers
    up.layer.stack.forEach((layer) => layer.mode !== "root" && layer.dismiss());
  }
});
```

**Reply by adam12** [reply]
The Glitch example helped. I'll look at it a bit later. :+1:

**Reply by hirasso** [reply]
Incidentally I came across exactly the same problem the last days. Basically, what I would expect unpoly to do if the back-button is being pressed with an open modal that has also triggered a URL change: 

1. Still have knowledge about the just opened modal
2. If the new URL is the same as the one before opening the modal, just close/remove the modal, do nothing else
3. If afterwards the forward-button is being pressed, re-open the modal

🤯😉

...by the way loving unpoly!! This is the only problem I have discovered so far.
